src/main/java/com/es/phoneshop/service/impl/HttpSessionCartService.java

```diff
@@ -0,0 +1,60 @@
+package com.es.phoneshop.service.impl;
+
+import com.es.phoneshop.dao.ProductDao;
+import com.es.phoneshop.dao.impl.ArrayListProductDao;
+import com.es.phoneshop.exception.OutOfStockException;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.cart.CartItem;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.service.CartService;
+
+import javax.servlet.http.HttpServletRequest;
+
+public class HttpSessionCartService implements CartService {
+    private static HttpSessionCartService instance;
+    private static final String CART_SESSION_ATTRIBUTE = HttpSessionCartService.class.getName() + ".cart";
+    private ProductDao productDao;
+
+    private HttpSessionCartService() {
+        productDao = ArrayListProductDao.getInstance();
+    }
+
+    public static synchronized HttpSessionCartService getInstance() {
+        if (instance == null) {
+            instance = new HttpSessionCartService();
+        }
+        return instance;
+    }
+
+    @Override
+    public synchronized Cart getCart(HttpServletRequest request) {
+        Cart cart = (Cart) request.getSession().getAttribute(CART_SESSION_ATTRIBUTE);
+        if (cart == null) {
+            cart = new Cart();
+            request.getSession().setAttribute(CART_SESSION_ATTRIBUTE, cart);
+        }
+        return cart;
+    }
+
+    @Override
+    public synchronized void add(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        Product product = productDao.getProduct(productId);
+        for (CartItem item : cart.getItems()) {
+            if (item.getProduct().getId().equals(productId)) {
+                int resultQuantity = item.getQuantity() + quantity;
+                int index = cart.getItems().indexOf(item);
+                checkStock(product, quantity, cart.getItems().get(index).getQuantity());
+                cart.getItems().get(index).setQuantity(resultQuantity);
```

Comment:
User should not be able to have negative amount of items in the cart. Though he should be able to add negative amount of products in the cart to remove some if he'd like.